xarray/core/indexes.py

```diff
@@ -100,3 +98,24 @@ def roll_index(index: pd.Index, count: int, axis: int = 0) -> pd.Index:
         return index[-count:].append(index[:-count])
     else:
         return index[:]
+
+
+def copy_indexes(
```

Comment:
Should this be a DataArray & Dataset method or just a utility function?

Reply:
Forgive me if this is naive (I haven't been following this as closely as others have), but why do we need to copy the indexes? They're immutable and (IIUC) the hash tables are not copied, so copying may mean we need to recreate the hash table after each copy.

Reply:
No you're not naive! I just don't know what I'm doing :)

I've removed the `.copy` below, so we are just moving the actual Index around. Basically I needed a utility function that could create new `_indexes` dictionaries excluding certain dimensions as needed and that accounted for `self._indexes = None`

What do you think would be a better name for this function? `propagate_indexes`?

Reply:
That makes two of us! :)

`propagate_indexes` is great, I think

Reply:
There's one issue about not copying indexes—while the index's values are immutable, its attributes (e.g. its name) aren't. As long as every other downstream operation doesn't mutate the index (for example it uses `idx.rename("foo")` rather than `idx.name = "foo"`, then we're fine. But potentially that's not robust.

@shoyer let us know if you have guidance!

Reply:
:+1: The current version does this:

```
new_indexes = {k: v for k, v in indexes.items() if k not in exclude}
```
which should be fine?